src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java

```diff
@@ -137,74 +146,120 @@ public String getScript() {
-        FilePath resultFile = c.getResultFile(ws);
-        FilePath controlDir = c.controlDir(ws);
-        if (capturingOutput) {
-            cmd = String.format("pid=$$; { while [ \\( -d /proc/$pid -o \\! -d /proc/$$ \\) -a -d '%s' -a \\! -f '%s' ]; do touch '%s'; sleep 3; done } & jsc=%s; %s=$jsc %s '%s' > '%s' 2> '%s'; echo $? > '%s.tmp'; mv '%s.tmp' '%s'; wait",
-                controlDir,
-                resultFile,
-                logFile,
-                cookieValue,
-                cookieVariable,
-                interpreter,
-                scriptPath,
-                c.getOutputFile(ws),
-                logFile,
-                resultFile, resultFile, resultFile);
-        } else {
-            cmd = String.format("pid=$$; { while [ \\( -d /proc/$pid -o \\! -d /proc/$$ \\) -a -d '%s' -a \\! -f '%s' ]; do touch '%s'; sleep 3; done } & jsc=%s; %s=$jsc %s '%s' > '%s' 2>&1; echo $? > '%s.tmp'; mv '%s.tmp' '%s'; wait",
-                controlDir,
-                resultFile,
-                logFile,
-                cookieValue,
-                cookieVariable,
-                interpreter,
-                scriptPath,
-                logFile,
-                resultFile, resultFile, resultFile);
-        }
-        cmd = cmd.replace("$", "$$"); // escape against EnvVars jobEnv in LocalLauncher.launch
+        envVars.put(cookieVariable, "please-do-not-kill-me");
 
-        if (LAUNCH_DIAGNOSTICS) {
-            args.addAll(Arrays.asList("sh", "-c", cmd));
+        String arch = ws.act(new getArchitecture());
+        List<String> launcherCmd = null;
+        String launcherBinary = LAUNCHER_PREFIX + os.toString() + arch;
+        InputStream launcherStream = DurableTask.class.getResourceAsStream(launcherBinary);
+        if ((launcherStream != null) && !FORCE_SHELL_WRAPPER) {
+            FilePath controlDir = c.controlDir(ws);
+            FilePath launcherAgent = controlDir.child(launcherBinary);
+            launcherAgent.copyFrom(launcherStream);
+            launcherAgent.chmod(0755);
+            launcherCmd = binaryLauncherCmd(c, ws, shell,
+                                            controlDir.getRemote(),
+                                            launcherAgent.getRemote(),
+                                            scriptPath,
+                                            cookieValue,
+                                            cookieVariable);
         } else {
-            // JENKINS-58290: launch in the background. Also close stdout/err so docker-exec and the like do not wait.
-            args.addAll(Arrays.asList("sh", "-c", "(" + cmd + ") >&- 2>&- &"));
+            String scriptString = scriptLauncherCmd(c, ws, interpreter, scriptPath, cookieValue, cookieVariable);
+            scriptString = scriptString.replace("$", "$$"); // escape against EnvVars jobEnv in LocalLauncher.launch
+            launcherCmd = new ArrayList<>();
+            if (LAUNCH_DIAGNOSTICS) {
```

Comment:
nit: I would put `scriptString = scriptString.replace("$", "$$")` and this if/else into `scriptLauncherCmd` since `binaryLauncherCmd` contains the equivalent.

Reply:
good catch :)

Reply:
Done in 4118e1a.